infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PreviewUrlCommandProvisioner.java

```diff
@@ -0,0 +1,115 @@
+ * {@link KubernetesNamespace}. When found, it composes full URL and set it as attribute of a
+ * command with key {@link
+ * org.eclipse.che.api.core.model.workspace.config.Command#PREVIEW_URL_ATTRIBUTE}.
+ *
+ * @param <E> type of the environment
+ */
+@Singleton
+public class PreviewUrlCommandProvisioner<E extends KubernetesEnvironment> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PreviewUrlCommandProvisioner.class);
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    injectsPreviewUrlToCommands(env, namespace);
+  }
+
+  /**
+   * Go through all commands, find matching service and exposed host. Then construct full preview
+   * url from this data and set it as Command's parameter under `previewUrl` key.
+   *
+   * @param env environment to get commands
+   * @param namespace current kubernetes namespace where we're looking for services and ingresses
+   */
+  private void injectsPreviewUrlToCommands(E env, KubernetesNamespace namespace)
+      throws InfrastructureException {
+    if (env.getCommands() == null) {
+      return;
+    }
+    for (CommandImpl command :
+        env.getCommands()
+            .stream()
+            .filter(c -> c.getPreviewUrl() != null)
+            .collect(Collectors.toList())) {
+      Optional<Service> foundService =
+          Services.findServiceWithPort(
+              namespace.services().get(), command.getPreviewUrl().getPort());
+      if (foundService.isPresent()) {
+        Optional<String> foundHost =
+            findHostForServicePort(
+                namespace, foundService.get(), command.getPreviewUrl().getPort());
+        if (foundHost.isPresent()) {
+          command.getAttributes().put(PREVIEW_URL_ATTRIBUTE, foundHost.get());
+        } else {
+          LOG.warn(
+              "unable to find ingress for service [{}] and port [{}]",
+              foundService.get(),
+              command.getPreviewUrl().getPort());
+        }
+      } else {
+        LOG.warn(
+            "unable to find service for port [{}] for command [{}]",
+            command.getPreviewUrl().getPort(),
+            command.getName());
+      }
+    }
+  }
+
+  protected Optional<String> findHostForServicePort(
+      KubernetesNamespace namespace, Service service, int port) throws InfrastructureException {
+    Optional<ServicePort> foundPort = Services.findPort(service, port);
+    if (!foundPort.isPresent()) {
+      return Optional.empty();
+    }
+
+    for (Ingress ingress : namespace.ingresses().get()) {
+      for (IngressRule rule : ingress.getSpec().getRules()) {
+        for (HTTPIngressPath path : rule.getHttp().getPaths()) {
+          IngressBackend backend = path.getBackend();
+          if (backend.getServiceName().equals(service.getMetadata().getName())
+              && backend.getServicePort().getStrVal().equals(foundPort.get().getName())) {
```

Comment:
What if service port is specified with int value? maybe it makes sense to compare kind
![Screenshot_20191021_175627](https://user-images.githubusercontent.com/5887312/67216610-2a7bb700-f42c-11e9-9264-01685bdbce16.png)

AFAIU 0 for int and 1 for string value.

Reply:
hmm, yes. And same in `OpenShift` one

Reply:
done